src/dlr_tensorflow/dlr_tensorflow.cc

```diff
@@ -0,0 +1,290 @@
+  std::string pb_file;
+  std::vector<std::string> paths_vec;
+  ListDir(dirname, paths_vec);
+  for (auto filename : paths_vec) {
+    std::string basename = GetBasename(filename);
+    if (EndsWith(filename, ".pb")) {
+      if (pb_file.empty()) {
+        pb_file = filename;
+      } else {
+        LOG(FATAL) << "Multiple .pb files under the folder: " << dirname;
+      }
+    }
+  }
+  if (pb_file.empty()) {
+    LOG(FATAL) << "No Tensorflow frozen model file found under folder: " << dirname;
+  }
+  return pb_file;
+}
+
+void dlr::FreeBuffer(void* data, size_t length) {
+  free(data);
+}
+
+TF_Buffer* dlr::ReadTFFile(const char* file) {
+  FILE *f = fopen(file, "rb");
+  fseek(f, 0L, SEEK_END);
+  long fsize = ftell(f);
+  fseek(f, 0L, SEEK_SET);  //same as rewind(f);
+
+  void* data = malloc(fsize);
+  if (fread(data, fsize, 1, f) != 1) {
+    printf("File read error....\n");
+  }
+  fclose(f);
+
+  TF_Buffer* buf = TF_NewBuffer();
+  buf->data = data;
+  buf->length = fsize;
+  buf->data_deallocator = FreeBuffer;
+  return buf;
+}
+
+void TensorflowModel::LoadFrozenModel(const char* pb_file) {
+  const char* op_prefix = "";
+  TF_ImportGraphDefOptions* opts = TF_NewImportGraphDefOptions();
+  TF_ImportGraphDefOptionsSetPrefix(opts, op_prefix);
+
+  TF_Buffer* graph_def = ReadTFFile(pb_file);
+  TF_GraphImportGraphDef(graph_, graph_def, opts, status_);
+  if (TF_GetCode(status_) != TF_OK) {
+    LOG(FATAL) << "ERROR: Unable to import graph " << TF_Message(status_);
+    return; // unreachable
+  }
+  LOG(INFO) << "Successfully imported graph";
+  TF_DeleteImportGraphDefOptions(opts);
+  TF_DeleteBuffer(graph_def);
+}
+
+void TensorflowModel::GenTensorSpec(bool isInput) {
+  std::vector<std::string> tensor_names = isInput ? input_names_ : output_names_;
+  std::regex r("^(.+):(\\d+)$");
+  for (std::string t_name : tensor_names) {
+    std::smatch match;
+    std::string op_name;
+    int op_out_id;
+    if (std::regex_search(t_name, match, r)) {
+      op_name = match.str(1);
+      op_out_id = std::stoi(match.str(2));
+    } else {
+      LOG(FATAL) << "ERROR: failed to parse tensor name " << t_name;
+      return; // unreachable
+    }
+
+    TF_Operation* op = TF_GraphOperationByName(graph_, op_name.c_str());
+    if (op == NULL) {
+      LOG(FATAL) << "ERROR: inputOp TF_GraphOperationByName failed for operation " << op_name;
+      return; // unreachable
+    }
+    TF_Output oper_out = {op, op_out_id};
+    int n_dim = TF_GraphGetTensorNumDims(graph_, oper_out, status_);
+    if (TF_GetCode(status_) != TF_OK) {
+      LOG(FATAL) << "ERROR: TF_GraphGetTensorNumDims failed " << TF_Message(status_);
+      return; // unreachable
+    }
+    int64_t dims[n_dim];
+    TF_GraphGetTensorShape(graph_, oper_out, dims, n_dim, status_);
+    if (TF_GetCode(status_) != TF_OK) {
+      LOG(FATAL) << "ERROR: TF_GraphGetTensorShape failed " << TF_Message(status_);
+      return; // unreachable
+    }
+    if (dims[0] == -1) {
```

Comment:
This will override the batch size to only allow batch size = 1 when the model has a dynamic batch dimension. If we want to allow dynamic shapes to work correctly, we need to defer `TF_AllocateTensor` until `SetInput` is called.

If we don't care about dynamic shapes and always want to override the batch dimension to 1 in those cases, we should change the dimension in the graph using `TF_GraphSetTensorShape`. This should allow TF to further optimize the graph by using static shapes.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h#L199-L219

Reply:
TVM compiled model has fixed batch size. DLR API was designed with this assumption.
Since compiled model has fixed batch size and DRL is designed for that we should ask about batch size in TensorflowModel constructor and call `TF_GraphSetTensorShape` in the initialization phase (before setInput)